lib/bpl/transformation/shadowing.rb

```diff
@@ -26,6 +26,8 @@ def shadow_copy(node)
         shadow
       end
 
+      def shadow_expr_eq(expr) "#{expr} == #{shadow_copy(expr)}" end
```

Comment:
Actually, attempting the merge, I now recall the actual strong typing
reason: shadow_eq is sometimes (l. 175, l. 189) called with a String
argument, which does not implement the copy method.

A solution to unify both cases would be to make sure we turn those strings
into variables on shadowing, but I'm not familiar enough with your classes
yet to figure out how.

F

On Fri, 18 Sep 2015 at 20:02 Michael Emmi <notifications@github.com> wrote:

> In lib/bpl/transformation/shadowing.rb
> <https://github.com/michael-emmi/bam-bam-boogieman/pull/3#discussion_r39883234>
> :
>
> > @@ -26,6 +26,8 @@ def shadow_copy(node)
> >          shadow
> >        end
> >
> > +      def shadow_expr_eq(expr) "#{expr} == #{shadow_copy(expr)}" end
>
> Not sure what you mean about “strong typing”: the Node-type argument of
> your new definition subsumes the Identifier-type argument of the previous
> definition.
>
> The shadow_copy function should do a full sub-expression traversal.
>
> If you unify your shadow_expr_eq into shadow_eq I will merge your pull
> request :-)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/michael-emmi/bam-bam-boogieman/pull/3/files#r39883234>
> .
>